### PR TITLE
feat(cta-section): web component version

### PR DESCRIPTION
**Pull request by IgnacioBecerra**
### Related Ticket(s)

#3749
#3368 

### Description

Introducing the `<dds-cta-section>` component for Web Components, featuring Footer variations with Content Items and a Link List.

Default:
Screen Shot 2020-11-04 at 3 34 01 PM

Content Items Footer:
Screen Shot 2020-11-04 at 3 34 15 PM

Link List Footer:
Screen Shot 2020-11-04 at 3 34 27 PM

### Changelog

**New**

- `<dds-cta-section>` component
- `<dds-cta-section-item>` component to wrap `<dds-cta-section-heading>` and `<dds-cta-section-copy>` with the correct styling

### CODE REVIEW

File: packages/web-components/src/components/cta-section/cta-section.ts [hunk 1]
```diff
@@ -23,6 +23,8 @@ const { stablePrefix: ddsPrefix } = ddsSettings;
  * The CTA Section pattern
  *
  * @element dds-cta-section
+ * @slot heading - The text heading.
+ * @slot buttons - The CTA Buttons.
  */
 @customElement(`${ddsPrefix}-cta-section`)
 class DDSCTASection extends StableSelectorMixin(LitElement) {
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/cta-section.ts:30]: Seems that this class should inherit `DDSContentItem`.

File: packages/web-components/src/components/cta-section/cta-section.scss [hunk 2]
```diff
@@ -13,11 +13,15 @@
...
+
+  ::slotted([slot='buttons']) {
+    margin-top: 1rem;
+  }
 }
 
 .#{$dds-prefix}--content-block__copy {
   display: inline-block;
-  font-size: 1.5rem;
+  font-size: calc(1.25rem + 0 * ((100vw - 20rem) / 62));
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/cta-section.scss:24]: Wondering where this change comes from...?

**Review by asudoh — COMMENTED** [review]
Nice progress @IgnacioBecerra!

File: packages/web-components/src/components/cta-section/cta-section.scss (hunk 2 above)
Comment (IgnacioBecerra) [reply packages/web-components/src/components/cta-section/cta-section.scss:24]: The font size change comes from the `React` version, I think it was changed from 1.5rem to that calculation recently. Comes from the `_cta-section.scss` under the `styles` package.

File: packages/web-components/src/components/cta-section/cta-section.ts (hunk 1 above)
Comment (IgnacioBecerra) [reply packages/web-components/src/components/cta-section/cta-section.ts:30]: Thanks for the heads up! I'm wondering though, what are the main reasons it should inherit `DDSContentItem`?

Comment (asudoh) [reply packages/web-components/src/components/cta-section/cta-section.ts:30]: Thank you for asking @IgnacioBecerra! One reason is that `<dds-cta-section>` seems to do a similar thing to `<dds-content-item>`, especially from the perspective of the API set. Another reason is that `<CTASection>` in React renders `<ContentItem>`, effectively extending `<ContentItem>` via [React composition pattern](https://reactjs.org/docs/composition-vs-inheritance.html), whereas React composition pattern doesn't work well with Web Components: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/00ad9c7/packages/web-components/docs/coding-conventions.md#preferring-class-inheritance-pattern-over-react-composition-pattern

File: packages/web-components/src/components/cta-section/__stories__/README.stories.mdx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import '@carbon/ibmdotcom-web-components/es/components/content-block/content-block-heading';
+import '@carbon/ibmdotcom-web-components/es/components/button-group/button-group';
+import '@carbon/ibmdotcom-web-components/es/components/button-group/button-group-item';
+
+```
+
+### HTML
+
+```html
+<dds-cta-section .copy="Lorem ipsum *dolor* sit amet">
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__stories__/README.stories.mdx:29]: Should use `<dds-content-item-copy>` for the copy content.

File: packages/web-components/examples/codesandbox/components/cta-section/index.html [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      body {
+        visibility: hidden;
+      }
+    </style>
+  </head>
+  <body>
+    <div class="bx--grid">
+      <div class="bx--row">
+        <div class="bx--col-sm-4 bx-col-lg-12 bx--offset-lg-4">
+          <dds-cta-section .copy="Lorem ipsum *dolor* sit amet">
```
Comment (asudoh) [inline packages/web-components/examples/codesandbox/components/cta-section/index.html:26]: Should use `<dds-content-item-copy>` for the copy content.

File: packages/web-components/src/components/cta-section/__stories__/cta-section.stories.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+const iconOptions = {
+  Default: null,
+  'Arrow Right': 'ArrowRight20',
+  'External Launch': 'Launch20',
+};
+
+export const Default = ({ parameters }) => {
+  const { heading, copy, renderIcon } = parameters?.props?.CTASection ?? {};
+  return html`
+    <dds-cta-section .copy="${ifNonNull(copy)}">
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__stories__/cta-section.stories.ts:37]: Should use `<dds-content-item-copy>` for the copy content.

File: packages/web-components/src/components/cta-section/__stories__/cta-section.stories.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        </dds-button-group-item>
+      </dds-button-group>
+    </dds-cta-section>
+  `;
+};
+
+export const WithContentItems = ({ parameters }) => {
+  const { heading, copy, renderIcon } = parameters?.props?.CTASection ?? {};
+  return html`
+    <dds-cta-section .copy="${ifNonNull(copy)}">
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__stories__/cta-section.stories.ts:55]: Should use `<dds-content-item-copy>` for the copy content.

File: packages/web-components/src/components/cta-section/__stories__/cta-section.stories.ts [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        <dds-text-cta slot="cta" cta-type="local" icon-placement="right" href="example.com">Browse tutorials</dds-text-cta>
+      </dds-cta-section-item>
+    </dds-cta-section>
+  `;
+};
+
+export const WithLinkList = ({ parameters }) => {
+  const { heading, copy, renderIcon } = parameters?.props?.CTASection ?? {};
+  return html`
+    <dds-cta-section .copy="${ifNonNull(copy)}">
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__stories__/cta-section.stories.ts:87]: Should use `<dds-content-item-copy>` for the copy content.

File: packages/web-components/src/components/cta-section/__tests__/cta-section.test.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+import { html, render } from 'lit-html';
+import ifNonNull from 'carbon-web-components/es/globals/directives/if-non-null.js';
+import '../cta-section';
+
+const template = (props?) => {
+  const { copy, children } = props ?? {};
+  return html`
+    <dds-cta-section .copy="${ifNonNull(copy)}">
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__tests__/cta-section.test.ts:17]: Should use `<dds-content-item-copy>` for the copy content.

**Review by asudoh — COMMENTED** [review]
Great progress - Thanks @IgnacioBecerra!

File: packages/web-components/src/components/cta-section/cta-section-item.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+ */
+@customElement(`${ddsPrefix}-cta-section-item`)
+class DDSCTASectionItem extends StableSelectorMixin(DDSContentItem) {
+  render() {
+    return html`
+      <div class="${prefix}--content-item">
+        <slot name="heading"></slot>
+        <slot></slot>
+        <slot name="cta"></slot>
+      </div>
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/cta-section-item.ts:33]: `<cta-section-item>` itself is a DOM element and thus rendering `<div class="bx--content-item">` will be a duplicate. We can use a tag selector to style `<cta-section-item>`: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.12.0/packages/web-components/docs/coding-conventions.md#custom-element-itself-as-an-eleement

```suggestion
      <slot name="heading"></slot>
      <slot></slot>
      <slot name="cta"></slot>
```

File: packages/web-components/src/components/cta-section/cta-section-item-heading.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+      this.setAttribute('aria-level', '4');
+    }
+    super.connectedCallback();
+  }
+
+  render() {
+    return html`
+      <div class="${prefix}--content-item__heading">
+        <slot></slot>
+      </div>
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/cta-section-item-heading.ts:47]: `<cta-section-item-heading>` itself is a DOM element and thus rendering `<div class="bx--content-item">` will be a duplicate. We can use a tag selector to style `<cta-section-item>`: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.12.0/packages/web-components/docs/coding-conventions.md#custom-element-itself-as-an-eleement

```suggestion
      <slot></slot>
```

File: packages/web-components/src/components/cta-section/cta-section-item-copy.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+const { stablePrefix: ddsPrefix } = ddsSettings;
+
+@customElement(`${ddsPrefix}-cta-section-item-copy`)
+class DDSCTASectionItemCopy extends StableSelectorMixin(DDSContentItemCopy) {
+  render() {
+    return html`
+      <div class="${prefix}--cta-section-item__copy">
+        <slot></slot>
+      </div>
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/cta-section-item-copy.ts:29]: `<cta-section-item-copy>` itself is a DOM element and thus rendering `<div class="bx--content-item">` will be a duplicate. We can use a tag selector to style `<cta-section-item>`: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.12.0/packages/web-components/docs/coding-conventions.md#custom-element-itself-as-an-eleement
```suggestion
      <slot></slot>
```

File: packages/web-components/src/components/cta-section/cta-section-copy.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+const { stablePrefix: ddsPrefix } = ddsSettings;
+
+@customElement(`${ddsPrefix}-cta-section-copy`)
+class DDSCTASectionItemCopy extends StableSelectorMixin(DDSContentItemCopy) {
+  render() {
+    return html`
+      <div class="${prefix}--cta-section__copy">
+        <slot></slot>
+      </div>
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/cta-section-copy.ts:28]: `<cta-section-copy>` itself is a DOM element and thus rendering `<div class="bx--content-item">` will be a duplicate. We can use a tag selector to style `<cta-section-item>`: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/blob/v1.12.0/packages/web-components/docs/coding-conventions.md#custom-element-itself-as-an-eleement
```suggestion
      <slot></slot>
```

File: packages/web-components/src/components/cta-section/cta-section-item.ts [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+  // eslint-disable-next-line class-methods-use-this
+  protected _renderContent() {
+    return html``;
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  protected _renderCopy() {
+    return '';
+  }
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/cta-section-item.ts:45]: Given `render()` doesn't use `_renderContent()`/`_renderCopy()` in this class, they seem to be redundant.

**Review by asudoh — COMMENTED** [review]
Thank you for the update @IgnacioBecerra!

**Comment by IgnacioBecerra** [comment]
@asudoh Thanks for the suggestions! Now using tag selection for all the component styling. :)

**Review by asudoh — APPROVED** [review]
LGTM 👍 - Thanks @IgnacioBecerra for all the updates!

**Review by kennylam — COMMENTED** [review]
@IgnacioBecerra Several components have had content removed (looks like markdown content) in the Percy snapshots. Any idea what's going on there?

**Comment by IgnacioBecerra** [comment]
@kennylam Yeah, I noticed that last night and was about comment about it. I'm using the new `<dds-content-item` and `<dds-content-item-copy` that @asudoh will be introducing in #4351. I didn't update the stories for all the components that use the component, thus the Percy snapshots. 

Since this PR is somewhat dependent on #4351, should we wait until that one goes through before merging this one?

**Comment by jeffchew** [comment]
Screen Shot 2020-11-06 at 11 12 54 AM

Minor thing, looks like the formatting for the docs is slightly off, can you take a look?

**Comment by kennylam** [comment]
> Since this PR is somewhat dependent on #4351, should we wait until that one goes through before merging this one, or should I edit the stories to reflect the changes?

@IgnacioBecerra Yes good call. That one looks ready to merge soon. We can update here and make sure all is good.

File: packages/web-components/src/components/cta-section/__stories__/README.stories.mdx [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+import '@carbon/ibmdotcom-web-components/es/components/button-group/button-group-item';
+
+```
+
+### HTML
+
+```html
+<dds-cta-section>
+  <dds-content-block-heading>Curabitur malesuada varius mi eu posuere</dds-content-block-heading>
+  <dds-cta-section-copy content="Lorem ipsum *dolor* sit amet"></dds-cta-section-copy>
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__stories__/README.stories.mdx:36]:
```suggestion
  <dds-cta-section-copy>Lorem ipsum *dolor* sit amet</dds-cta-section-copy>
```

File: packages/web-components/src/components/cta-section/__stories__/README.stories.mdx [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+      Primary button
+    </dds-button-group-item>
+  </dds-button-group>
+</dds-cta-section>
+```
+
+## Props
+
+<Props of="dds-cta-section" />
+
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__stories__/README.stories.mdx:51]:
```suggestion

<Description markdown={markdownContents} />
```

File: packages/web-components/src/components/cta-section/__stories__/README.stories.mdx [hunk 16]
```diff
@@ -0,0 +1,60 @@
+import { Props, Description } from '@storybook/addon-docs/blocks';
+import contributing from '../../../../../../docs/contributing-license.md';
+
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__stories__/README.stories.mdx:3]:
```suggestion
import markdownContents from '../../../../docs/markdown-contents.md';

```

File: packages/web-components/examples/codesandbox/components/cta-section/index.html [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+      }
+    </style>
+  </head>
+  <body>
+    <div class="bx--grid">
+      <div class="bx--row">"Lorem ipsum *dolor* sit amet"
+        <div class="bx--col-sm-4 bx-col-lg-12 bx--offset-lg-4">
+          <dds-cta-section>
+            <dds-content-block-heading>Curabitur malesuada varius mi eu posuere</dds-content-block-heading>
+            <dds-cta-section-copy content=${ifNonNull(copy)}></dds-cta-section-copy>
```
Comment (asudoh) [inline packages/web-components/examples/codesandbox/components/cta-section/index.html:28]:
```suggestion
            <dds-cta-section-copy>Want to discuss your options with a DevOps expert? Contact our sales team to evaluate your needs.</dds-cta-section-copy>
```

File: packages/web-components/src/components/cta-section/__stories__/cta-section.stories.ts [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+        <dds-button-group-item href="https://example.com">
+          Primary button ${renderIcon}
+        </dds-button-group-item>
+      </dds-button-group>
+
+      <dds-cta-section-item slot="content-item">
+        <dds-cta-section-item-heading>Get connected</dds-cta-section-item-heading>
+        <dds-cta-section-item-copy
+          >IBM DevOps partners have a wide range of expertise. Find one to build the right solution for you."
+        </dds-cta-section-item-copy>
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__stories__/cta-section.stories.ts:74]:
```suggestion
        <dds-cta-section-item-copy
          >IBM DevOps partners have a wide range of expertise. Find one to build the right solution for you.</dds-cta-section-item-copy>
```

File: packages/web-components/src/components/cta-section/__tests__/cta-section.test.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+import { html, render } from 'lit-html';
+import ifNonNull from 'carbon-web-components/es/globals/directives/if-non-null.js';
+import '../cta-section';
+
+const template = (props?) => {
+  const { copy, heading, children } = props ?? {};
+  return html`
+    <dds-cta-section>
+      <dds-content-block-heading>${ifNonNull(heading)}</dds-content-block-heading>
+      <dds-cta-section-copy content=${ifNonNull(copy)}></dds-cta-section-copy>
```
Comment (asudoh) [inline packages/web-components/src/components/cta-section/__tests__/cta-section.test.ts:19]:
```suggestion
      <dds-cta-section-copy>ifNonNull(copy)</dds-cta-section-copy>
```

**Review by asudoh — COMMENTED** [review]
Thank you @IgnacioBecerra for the updates and rebasing!

**Comment by IgnacioBecerra** [comment]
@asudoh Thanks for catching those! Updated with new component usage.

**Comment by asudoh** [comment]
Cool thanks @IgnacioBecerra for the update!

@jeffchew @kennylam Do you this is ready to go? Thanks!

**Review by jeffchew — APPROVED** [review]
LGTM!

**Pull request merged**